src/cmd/wallet.rs

```diff
@@ -89,11 +89,68 @@ pub enum AddressOpt {
     /// First one that hasn't been used.
     LastUnused,
     /// List addresses
-    List,
+    List {
+        /// Show all addresses (internal & external)
+        #[structopt(long, short)]
+        all: bool,
+        /// Show only internal addresses
+        #[structopt(long, short)]
+        internal: bool,
+        /// Hide addresses with zero balance
+        #[structopt(long)]
+        hide_zeros: bool,
+    },
     /// Show details of an address
     Show { address: Address },
 }
 
+fn list_keychain_addresses(
+    wallet_dir: &Path,
+    config: Config,
+    keychain_kind: KeychainKind,
+    hide_zeros: bool,
+) -> anyhow::Result<Vec<Vec<Cell>>> {
+    let wallet_db = load_wallet_db(wallet_dir).context("loading wallet db")?;
+    let scripts = wallet_db.iter_script_pubkeys(Some(keychain_kind))?;
+    let index = wallet_db.get_last_index(keychain_kind)?;
+    let map = index_utxos(&wallet_db)?;
+    let rows = match index {
+        Some(index) => scripts
+            .iter()
+            .take(index as usize + 1)
+            .map(|script| {
+                let address = Address::from_script(script, config.network).unwrap();
+                let value = map
+                    .get(script)
+                    .map(|utxos| Amount::from_sat(utxos.iter().map(|utxo| utxo.txout.value).sum()))
+                    .unwrap_or(Amount::ZERO);
+
+                if hide_zeros && (value == Amount::ZERO) {
+                    return vec![];
+                }
+
+                let count = map.get(script).map(Vec::len).unwrap_or(0);
+                let keychain_name = if let KeychainKind::External = keychain_kind {
+                    "external"
+                } else {
+                    "internal"
+                }
+                .to_string();
```

Comment:
:bulb: more rusty approach:

```suggestion
                let keychain_name = match keychain_kind {
                    KeychainKind::External => "external",
                    KeychainKind::Internal => "internal"
                }.to_string();
```